src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdg_decoration_unstable_v1.h"
+
+#include "mir/shell/surface_specification.h"
+#include "mir/wayland/protocol_error.h"
+
+#include "xdg-decoration-unstable-v1_wrapper.h"
+#include "xdg_output_v1.h"
+#include "xdg_shell_stable.h"
+
+#include <memory>
+#include <unordered_set>
+
+namespace mir
+{
+namespace frontend
+{
+class XdgDecorationManagerV1 : public wayland::XdgDecorationManagerV1
+{
+public:
+    XdgDecorationManagerV1(wl_resource* resource);
+
+    class Global : public wayland::XdgDecorationManagerV1::Global
+    {
+    public:
+        Global(wl_display* display);
+
+    private:
+        void bind(wl_resource* new_zxdg_decoration_manager_v1) override;
+    };
+
+private:
+    void get_toplevel_decoration(wl_resource* id, wl_resource* toplevel) override;
+	std::shared_ptr<std::unordered_set<wl_resource*>> toplevels_with_decorations;
```

Comment:
I think that rather than sharing a "naked" set, you should wrap it in a class with `register/unregister` methods. That makes the invariants a lot clearer.

Reply:
Done. I'm not so sure about the naming, but I don't think it's a big deal since it's something internal.